File: src/main/cpp/body.cpp
#include "body.hpp"

#include <stdio.h>
#include <cstdio>
#include <iostream>
#include <fstream>
#include <string>
#include <iomanip>
#include <vector>
#include <string>

using namespace std;

body::~body() {
}

body::body() {
    nom = "";
    m = 0;
    mg = 0;

    rx = 0;
    ry = 0;
    rz = 0;

    vx = 0;
    vy = 0;
    vz = 0;

    ax = 0;
    ay = 0;
    az = 0;
}

body::body(Json::Value JsonBody) {
    this->nom = JsonBody["_name"].asString();

    this->m = JsonBody["m"].asDouble();
    this->mg = m * G;

    this->rx = JsonBody["rx"].asDouble();
    this->ry = JsonBody["ry"].asDouble();
    this->rz = JsonBody["rz"].asDouble();

    this->vx = JsonBody["vx"].asDouble();
    this->vy = JsonBody["vy"].asDouble();
    this->vz = JsonBody["vz"].asDouble();
}

body::body(vector<string> csv) {
    string::size_type sz;
    this->nom = csv[0];
    this->m = stod(csv[1]);
    this->mg = m * G;

    this->rx = stod(csv[2]);
    this->ry = stod(csv[3]);
    this->rz = stod(csv[4]);

    this->vx = stod(csv[5]);
    this->vy = stod(csv[6]);
    this->vz = stod(csv[7]);
}

Json::Value body::toJson() {
    Json::Value root;
    root["_name"] = this->nom;
    root["m"] = this->m;
    root["rx"] = this->rx;
    root["ry"] = this->ry;
    root["rz"] = this->rz;
    root["vx"] = this->vx;
    root["vy"] = this->vy;
    root["vz"] = this->vz;

    return root;
}

Json::Value body::toJsonLight() {
    Json::Value root;
    root["id"] = this->id;
    root["rx"] = this->rx;
    root["ry"] = this->ry;
    root["rz"] = this->rz;

    return root;
}

void body::actualise() {
    ax /= mg;
    ay /= mg;
    az /= mg;
    // Calcul de la nouvelle position
    rx += 0.5 * ax * interval_p2 + vx * interval;
    ry += 0.5 * ay * interval_p2 + vy * interval;
    rz += 0.5 * az * interval_p2 + vz * interval;
    // Calcul de la nouvelle vitesse
    vx += ax * interval;
    vy += ay * interval;
    vz += az * interval;
}

File: src/main/rust/body.rs
#[derive(Debug, Clone)]
pub struct Body {
    pub id: u32,
    pub name: String,
    pub m: f64,
    // position
    pub rx: f64,
    pub ry: f64,
    pub rz: f64,
    // velocity
    pub vx: f64,
    pub vy: f64,
    pub vz: f64,
    // acceleration
    pub ax: f64,
    pub ay: f64,
    pub az: f64,
}

impl Body {
    pub fn update_pos(&mut self, duration_sec: f64) {
        self.ax /= self.m;
        self.ay /= self.m;
        self.az /= self.m;

        self.rx += 0.5 * self.ax * duration_sec * duration_sec + self.vx * duration_sec;
        self.ry += 0.5 * self.ay * duration_sec * duration_sec + self.vy * duration_sec;
        self.rz += 0.5 * self.az * duration_sec * duration_sec + self.vz * duration_sec;

        // Calcul de la nouvelle vitesse
        self.vx += self.ax * duration_sec;
        self.vy += self.ay * duration_sec;
        self.vz += self.az * duration_sec;

        self.ax = 0.0;
        self.ay = 0.0;
        self.az = 0.0;
    }
}

pub struct Frame {
    pub id: u32,
    pub bodies: Vec<Body>,
}
